lib/logstash/outputs/elasticsearch/data_stream_support.rb

```diff
@@ -0,0 +1,247 @@
+      # Defaults to `false` in Logstash 7.x and `auto` starting in Logstash 8.0.
+      base.config :data_stream, :validate => ['true', 'false', 'auto']
+
+      base.config :data_stream_type, :validate => ['logs', 'metrics', 'synthetics'], :default => 'logs'
+      base.config :data_stream_dataset, :validate => :dataset_identifier, :default => 'generic'
+      base.config :data_stream_namespace, :validate => :namespace_identifier, :default => 'default'
+
+      base.config :data_stream_sync_fields, :validate => :boolean, :default => true
+      base.config :data_stream_auto_routing, :validate => :boolean, :default => true
+
+      base.extend(Validator)
+    end
+
+    # @override
+    def finish_register
+      super
+
+      if data_stream_config?
+        @event_mapper = -> (e) { data_stream_event_action_tuple(e) }
+        @event_target = -> (e) { data_stream_name(e) }
+        @index = "#{data_stream_type}-#{data_stream_dataset}-#{data_stream_namespace}".freeze # default name
+      end
+    end
+
+    # @note assumes to be running AFTER {after_successful_connection} completed, due ES version checks
+    def data_stream_config?
+      @data_stream_config.nil? ? @data_stream_config = check_data_stream_config! : @data_stream_config
+    end
+
+    private
+
+    def data_stream_name(event)
+      data_stream = event.get('data_stream')
+      return @index if !data_stream_auto_routing || !data_stream.is_a?(Hash)
+
+      type = data_stream['type'] || data_stream_type
+      dataset = data_stream['dataset'] || data_stream_dataset
+      namespace = data_stream['namespace'] || data_stream_namespace
+      "#{type}-#{dataset}-#{namespace}"
+    end
+
+    # @param params the user configuration for the ES output
+    # @note LS initialized configuration (with filled defaults) won't detect as data-stream
+    # compatible, only explicit (`original_params`) config should be tested.
+    # @return [TrueClass|FalseClass] whether given configuration is data-stream compatible
+    def check_data_stream_config!(params = original_params)
+      use_data_stream = data_stream_explicit_value
+      data_stream_params = params.select { |name, _| name.start_with?('data_stream_') } # exclude data_stream =>
+      invalid_data_stream_params = invalid_data_stream_params(params)
+
+      if use_data_stream.eql?(false) && data_stream_params.any?
+        @logger.warn "Ignoring data stream specific configuration (due data_stream => false)", data_stream_params
+      end
+
+      if use_data_stream.nil?
+        use_data_stream = data_stream_default(invalid_data_stream_params.empty?)
+        if !use_data_stream && data_stream_params.any?
+          # DS (auto) disabled but there's still some data-stream parameters (and no `data_stream => false`)
+          @logger.warn "Ambiguous configuration, data stream settings have no effect", data_stream_params
+          raise LogStash::ConfigurationError, "Ambiguous configuration, please set data_stream => true/false " +
```

Comment:
If this is run in the `finish_register` hook in another thread, I don't think we can guarantee that we properly cascade the error and prevent the pipeline from starting. For example, if there is some latency, `Plugin#register` could have already returned before a connection is established and the framework we have for dealing with `LogStash::ConfigurationError` would have been bypassed and we will instead crash the output after it has possibly begun processing events.

We _may_ need to flip things around and block in `register` until the `finish_register` has succeeded.

The other option I can see is to _not_ rely on the state of Elasticsearch and to simply default to using datastreams in Logstash 8 regardless of what we are connected to, possibly guarded by whether Logstash itself is Licensed or OSS-only.

Reply:
Good point, we could retrieve the error on the LS worker thread but at this point I would rather try:
> We may need to flip things around and block in register until the finish_register has succeeded.

Reply:
Looked into this and waiting for a successful connection in `register` is problematic - by blocking we effectively disable the pipeline setup. Now we could try to be smart with a timeout but it's making things even more complicated.

A reasonable resolution is for the DS setup to **NOT** rely on ES version check and effectively always default to `data_stream => false` (assuming no `data_stream...` configuration present) on LS 7.x, potentially with an info message that the default will change in LS 8.0. I think the attempt to play things smart wasn't a good idea ...